### ISSUE TITLE
Ultimate backporting tool

### ISSUE THREAD

**Issue by annulen** [annulen/webkit#551]
* Scan trunk from branching point (revision number) to the future
    * Maybe the best way is to use Trac, maybe it has some API. It's possible to use mirrored git braches, but we need to maintain them somehow. But it's possible to automate that too.
* Hide revisions that we already have in stable branch
    * Enhancement: hide revisions that we already sorted out, results can be saved in git notes, like in [1,2]
* Hide revisions that don't affect any file from Tools/qt/manifest.txt
* Indicate if revision is backported to some branch of Apple or GTK
* Indicate if bugzilla url of the change has restricted access
* (Advanced) Indicate if patch is reverted in any downstream branch
* (Maybe overkill) Indicate if revision has topic REGRESSION(rN), and we have rN in branch

[1] http://blog.adamspiers.org/2013/09/19/easier-upstreaming-with-git/
[2] http://blog.adamspiers.org/2013/10/02/more-uses-for-git-notes/

**Comment by annulen** [comment]
Wishlist: categorize patches, e.g. currently I'm skipping all AX and Inspector patches

**Comment by annulen** [comment]
I've successfully used following manual process:
1. Open branch in Trac, e.g. https://trac.webkit.org/browser/webkit/branches/safari-602-branch
2. Copy list of revisions from browser to text file
3. Reverse it with `tac` command so older revisions are in the beginning
4. List 10 entries with `head`
5. get revison number without "r" from commit message, use `git log --since=2016 --grep=$number` and  `git log --since=2016 --grep=$number origin/master`
6. If I need this change, work with it
7. Go to 5 and take next line
8. When 10 entries are done, remove 10 lines from file and go to 4

There is one benefit in using git log --grep: I see all later revisions that refer to give one, so I can easily spot fixups and regressions

**Comment by annulen** [comment]
Of course, manual process won't be efficient when screening changes from trunk, as there will be much more revisions to process

**Comment by annulen** [comment]
Read https://trac.webkit.org/wiki/QtWebKitReleases and https://trac.webkit.org/wiki/QtWebKitSecurity

There is a tool cherry-pick-into-release-branch.py in http://gitorious.org/qtwebkit/tools.git repo, maybe it can be partially reused

**Comment by annulen** [comment]
@ashkulz have prepared scripts which automatically export SVN branches to git, you can see result in https://github.com/qtwebkit/webkit. So we don't need to use Trac to implement this tool and can get all information from querying git branches

**Comment by ashkulz** [comment]
I'll be pushing the script soon, had a medical procedure over the weekend.

**Comment by annulen** [comment]
Thanks! No hurry

**Comment by ashkulz** [comment]
@annulen do you want only the WebkitGTK branches or the Safari branches or both? It requires slightly different configuration for each case.

**Comment by annulen** [comment]
Both, but for Safari only branches named safari-<integer>-branch are interesting, like safari-607-branch.

**Comment by ashkulz** [comment]
Okay, I'm re-running it from scratch in a new clone (which is taking a lot of time!). Do you want me to submit a PR with it in `Tools/Scripts` or it'll go into a separate repo which has tools you'll be using for merging/backports?

**Comment by annulen** [comment]
For now let's use a separate repo. It may to go into `Tools/Scripts` later if upstream is interested, but if it's used only by us I'd prefer to have it outside

**Comment by annulen** [comment]
BTW, will new clone be required each time new branch is added?

**Comment by ashkulz** [comment]
Nope, it's a one-time thing (_you have to go through svn history to find branch points_). Updating it does take longer though, if branches have been created/updated (max ~5 minutes, most cases).

**Comment by annulen** [comment]
To save time you can import only the most recent of gtk and safari branches. Most important for us are branches which will be created in the next 2 months

**Comment by ashkulz** [comment]
> For now let's use a separate repo.

Which one? Or just create a gist?

> To save time you can import only the most recent of gtk and safari branches. Most important for us are branches which will be created in the next 2 months

Cool, do you want me to push them to this repo? I think we can delete the existing qtwebkit/webkit repository if we're doing that.

**Comment by annulen** [comment]
>Which one? Or just create a gist?

You can name it e.g. "mirror-scripts" or smth alike. Gist is not a good option

>I think we can delete the existing qtwebkit/webkit repository if we're doing that.

Yeah, it would be better to create new repo specially for mirrors

**Comment by annulen** [comment]
I'm using following workflows for backporting

### Backport patches which other ports have backported

1. Find stable branch with branching points near mine. For example, if I'm backporting to qtwebkit/qtwebkit@qtwebkit-5.212, near branches are qtwebkit/webkit-mirror@webkitgtk/webkit-2.12, qtwebkit/webkit-mirror@safari/safari-602-branch, and to some extent qtwebkit/webkit-mirror@webkitgtk/webkit-2.14
2. Look though all commits (I'll call them "candidates") in that branch for which following conditions are true:
* Candidate is a commit backported from trunk, not something original
* Candidate should affect at least one of files used by us, i.e. match one of patterns from https://raw.githubusercontent.com/qtwebkit/qtwebkit/qtwebkit-dev/Tools/qt/manifest.txt. )Counter-example: some patch affects a group of .mm files which are Apple-specific and missing in manifest, such patch is not a candidate)
* Candidate was not reverted in that branch later after backporting. (If some commit is backported, than reverted, than backported again, it counts as candidate)
3. Try to cherry-pick found candidates

### Backport patches which affect certain files
1. Look through some range of git log on master for given files
2. Candidates are patches which were not reverted later in history
